src/ddmd/expressionsem.d

```diff
@@ -6822,7 +6829,10 @@ extern (C++) final class ExpressionSemanticVisitor : Visitor
         }
 
         exp.type = exp.e1.type;
-        result = exp.reorderSettingAAElem(sc);
+        auto res = exp.reorderSettingAAElem(sc);
+        if (!appendArray)
```

Comment:
Why isn't the check done for `T[] ~= T[]` ?
If `T` is a reference type, then the assignment should be checked as well. I checked out your branch and tested it:
```D
void main () @safe
{
    Object o = test();
    assert(o !is null);
}

Object test() @safe
{
    scope Object obj = new Object;
    scope Object[] arr;
    arr ~= obj;
    Object[] array;
    array ~= arr;
    return array[0];
}
```

It compiles and pass.

Reply:
I believe that requires distinctly different code to implement, and so in order to avoid confusing things, should be a separate bugzilla issue and PR.

Reply:
https://issues.dlang.org/show_bug.cgi?id=17842

Reply:
Somehow I forgot to reply here...
The reason I came up with this test-case was because the new code is explicitly checking for it, and explicitly excluding it. But if you prefer to separate the issue, fine by me.

Reply:
There are 3 completely distinct cases for appending:

1. appending an array to an array
2. appending an element to an array
3. appending a multibyte utf character to an array

This fixes 2. Your case is 1. After this is pulled, I intend to do a bit of refactoring to make that clearer, then attend to case 1.

Reply:
Perfect.
However, as you can see from the failing Jenkins test, this check should be wrapped to only be performed with `-dip1000` or it'll break code.

Reply:
ok